docs/developer/architecture/control-plane-ingress.rst

```diff
@@ -0,0 +1,164 @@
+
+MetalLB is not deployed in every environment so it needs to be configurable
+from the Bootstrap configuration file, that’s why we have a new field about
+MetalLB in the Control Plane network section.
+
+MetalLB can use 2 different protocols either Layer2 either BGP, by default
+we will use a simple Layer2 configuration for MetalLB so that if the user
+environment allows using Layer2 protocol of MetalLB you just have to enable
+MetalLB, provide an Ingress IP for the Control Plane and that it.
+
+Default MetalLB configuration:
+
+.. code-block:: yaml
+
+    address-pools:
+    - name: ingress-ip
+      protocol: layer2
+      addresses:
+      - <control-plane ingress ip>/32
+      auto-assign: false
+
+But, since BGP router configuration can be really different between various
+environments, we need to make sure this
+`MetalLB configuration <https://metallb.universe.tf/configuration/#bgp-configuration>`_
+is exposed, that’s why we also have the ability to override default MetalLB
+configuration from the Bootstrap configuration file in order to provide BGP
+router information.
+
+Same as the Control Plane Ingress IP, we can switch from non-MetalLB to MetalLB
+(and the opposite) at any time just by following the same procedure.
+
+Deployment
+~~~~~~~~~~
+
+As for every other addon in MetalK8s, we will use the MetalLB helm chart and
+render this one using a specific “option” file. But this one will not be always
+deployed as we only want to deploy it when a specific key is set in the
+Bootstrap configuration file, so in the Salt pillar at the end.
+
+When we use MetalLB we do not want to use the same NGINX Ingress Controller
+deployments, since MetalLB will be the entry point in the Kubernetes cluster
+we do not need to use a DaemonSet running on every Control Plane nodes,
+instead, we will use a Deployment with 2 replicas.
+
+We also need to configure the Service for Ingress Controller differently
+depending on if we use MetalLB or not when we use it we want to use a
+LoadBalancer service and set the LoadBalancerIP to IngressIP provided by
+the user. If we do not use MetalLB then we use ClusterIP Service with IngressIP
+provided by the user as External IPs.
+
+It means the deployment of NGINX Ingress Controller depends on some Salt
+pillar values, also since we want to be able to switch between MetalLB and
+non-MetalLB we need to make sure the Salt states that deploy NGINX Ingress
+Controller remove no-longer-needed objects (e.g.: if you switch from
+non-MetalLB to MetalLB you want to remove the DaemonSet for NGINX Ingress
+Controller).
+
+Documentation
+-------------
+
+- Describe all new Bootstrap configuration fields (with some links
+  to MetalLB documentation to configure MetalLB to use BGP protocol)
+- Add a simple procedure to change the Control Plane Ingress IP and
+  reconfigure all Control Plane conponents that need to.
+
+Test Plan
+---------
+
+Add some End-to-End tests in the CI:
+
+- Use MetalLB and a VIP as Control Plane Ingress IP
```

Comment:
Likely needs a VIP failover test scenario as well.

Reply:
Hmmm, right but not sure it's really easy to do, I don't know if we have an easy way to know where this VIP sit at any point in time (except doing an ARP request to get mac address and then checking the host having the network interface with this mac address)
Anyway I agree, I will add it here

Reply:
(Added)